UserDB.cpp

```diff
@@ -80,10 +131,152 @@ void UserDB::Stop()
         lock.unlock();
         _inotify_thread.join();
         _update_thread.join();
+
+        if (listener_thread.joinable()) {
+            listener_thread.join();
+        }
+        sd_bus_unref(bus);
+        bus = nullptr;
+    }
+}
+
+void UserDB::ListenForUserChanges() {
+    // Add match rules for user added and removed signals
+
+    sd_bus_match_signal(bus, nullptr, "org.freedesktop.login1", "/org/freedesktop/login1", "org.freedesktop.login1.Manager", "UserNew", user_added_handler, this);
+    sd_bus_match_signal(bus, nullptr, "org.freedesktop.login1", "/org/freedesktop/login1", "org.freedesktop.login1.Manager", "UserRemoved", user_removed_handler, this);
+
+    while (!_stop) {
+        int ret = sd_bus_process(bus, nullptr);
+        if (ret < 0) {
+            Logger::Error("Failed to process bus: %s", strerror(-ret));
+            break;
+        }
+        if (ret > 0) {
+            continue; // Continue to process any further pending events
+        }
+
+        // Wait for the next event
+        ret = sd_bus_wait(bus, UINT64_MAX);
+        if (ret < 0) {
+            Logger::Error("Failed to wait on bus: %s", strerror(-ret));
+            break;
+        }
+    }
+}
+
+int UserDB::user_added_handler(sd_bus_message* m, void* userdata, sd_bus_error*) {
+    UserDB* db = static_cast<UserDB*>(userdata);
+    uint32_t uid;
+    const char* username;
+
+    int ret = sd_bus_message_read(m, "us", &uid, &username);
+    if (ret < 0) {
+        Logger::Error("Failed to parse UserNew signal: %s", strerror(-ret));
+        return ret;
+    }
+
+    db->user_map[uid] = username;
+    return 0;
+}
+
+int UserDB::user_removed_handler(sd_bus_message* m, void* userdata, sd_bus_error*) {
```

Comment:
>sd_bus_error [](http://example.com/codeflow?start=68&length=12)

Could you clarify the purpose?

Reply:
This is a required parameter for the callback's signature (sd_bus_message_handler_t) on the listener. Reference - https://www.freedesktop.org/software/systemd/man/latest/sd_bus_add_match.html#:~:text=sd_bus_add_match%20()%20installs%20a%20match%20rule%20for%20messages%20received%20on